<h1>Python Programming: Variables</h1>

## Overview
This lesson introduces students to the concept of creating and assigning variables - numeric variables (including basic math functions / numerical operators) and string variables (including concatenation). There is a small section near the end on different ways to print these variables.

---
## Lesson Objectives
*After this lesson, you will be able to...*

* Create and re-assign numerical and string variables.
* Use numerical operators.
* Print complex variable structures.

---
## What's a Variable?

Turn to the person next to you, and together come up with as many definitions for the word "variable" as you can.

- Consider contexts such as mathematics, the sciences, weather, etc.
- No cheating! Phones off and laptops closed.
---
## Variable


Variables:

- Are boxes that can hold all kinds of information for you.
- Make it easier to store and re-use values.
- Are the most basic piece of code.

To use a variable, we simply announce that we want to use it (we **declare** it).



In [2]:
# I've eaten 3 cupcakes
cupcakes_ive_eaten = 3
print(cupcakes_ive_eaten)
# Prints 3


3


---
## Naming Conventions: Mistakes and Syntax

Some common naming mistakes:

- Not using meaningful names. `delicious = 3` doesn't mean anything - `cupcakes_ive_eaten = 3` does!
- Case sensitivity (`CUPCAKES_IVE_EATEN` and `cupcakes_ive_eaten` are not the same!)
- No spaces or punctuation ("cupcakes i've eaten" isn't allowed)
  - This is invalid **syntax**
  - Use snake_case: `lowercase_letters_with_underscores` (it's in the official [Python style guide](https://www.python.org/dev/peps/pep-0008/#descriptive-naming-styles))
---
## Discussion: Changing Values

What if, later, you eat more cupcakes? Now, this is wrong.

In [3]:
cupcakes_ive_eaten = 3

---
## Discussion: Reassigning Variables



In the example below, what do you think the output of the code is?



In [4]:
cupcakes_ive_eaten = 3
print(cupcakes_ive_eaten)
cupcakes_ive_eaten = 4
print(cupcakes_ive_eaten)

3
4


---
## Quick Review

- A variable is a box that holds a value.

- It can be declared, called, and changed within your program.

- When declaring variables, syntax and naming conventions matter!

- Variables can be reassigned as often as you like, but only the most recent declaration counts.

**UP NEXT:** Math!

---
## Mathematical Operators

Math works on numerical variables, too!

- The `+`, `-`, `*` (multiply), and `/` (divide) operators work just like they do with regular math.



In [5]:
cupcakes_ive_eaten = 6 + 3
print(cupcakes_ive_eaten)
# Prints 9

cupcakes_ive_eaten = 6 - 3
print(cupcakes_ive_eaten)
# Prints 3

cupcakes_ive_eaten = 6 * 3
print(cupcakes_ive_eaten)
# Prints 18

cupcakes_ive_eaten = 6 / 3
print(cupcakes_ive_eaten)
# Prints 2


9
3
18
2.0


---
## Even More Mathematical Operators

Beyond the `+`, `-`, `*` (multiply), and `/` (divide) operators, we have modulus and exponents.


In [6]:
making_exponents = 10 ** 2
print(making_exponents)
# Prints 100

more_exponents = 10 ** 3
print(more_exponents)
# Prints 1,000

making_modulus = 10 % 3
print(making_modulus)
# Prints 1

more_modulus = 6 % 2
print(more_modulus)
# Prints 0


100
1000
1
0


---
## Math On The Same Variable


You can reassign a variable *using that very same variable* - or other variables!



In [10]:
cupcakes_ive_eaten = 3
cupcakes_ive_eaten = cupcakes_ive_eaten + 1
print(cupcakes_ive_eaten)
# Prints 4.

cupcakes_left_in_box = 6
cupcakes_left_in_box = cupcakes_left_in_box - 1 
print(cupcakes_left_in_box)
# Prints 5.

cupcakes_left_in_box = cupcakes_left_in_box - cupcakes_ive_eaten 
print(cupcakes_left_in_box)
# Prints 1.


4
5
1


---
## Partner Exercise: Mathematical Operators

In pair programming, one person serves as the driver -- they are sharing their screen and typing -- and one person serves as the navigator -- they provide guidance as the code is written.
- When you pair up, decide who will be the driver and who will be the navigator. Then, follow the directions in the comments.

In [11]:
# Make a variable to hold the number of guitars you own (3).
guitars_i_own = 3

# Make a variable to hold the number of guitars Nikhil owns (1).
guitars_nikhil_owns = 1

# You give 2 of your guitars to Nikhil, so subtract 2 from you and add 2 to Nikhil.
guitars_i_own = guitars_i_own - 2
guitars_nikhil_owns = guitars_nikhil_owns + 2

print(guitars_i_own)
print(guitars_nikhil_owns)

1
3


---
## Reassignment Shorthand

This is okay:

In [14]:
my_num = 9
my_num = my_num + 7
# my_num is now 16
print(my_num)

16



But this is better:



In [15]:
my_num = 9
my_num += 7 # += is short for theSameVariable = theSameVariable + 7
# my_num is now 16
print(my_num)

16


This works with `+=`, `-=`, `*=`, `/=` - any math operations.

---
## Partner Exercise: Numerical Reassignment

Get with the same partner, but switch driver and navigator roles.

- Remember, in pair programming, one person serves as the driver -- their hands are on the keyboard -- and one person serves as the navigator -- they provide guidance as the code is written.
- Follow the directions in the comments.

In [16]:
# - Declare two variables `num1` and `num2` and assign them to any numbers you'd like.
num1 = 5
num2 = 9
# - Set `num1` to the result of subtracting `num1` from the `num2`.
num1 = num2 - num1
# - Create a new variable `num3` that will help us tell if `num2` is even or odd.
num3 = num2 % 2
# - Using shorthand, add 5 to `num1`.
num1 += 5
# - Print out `num1`, `num2`, and `num3`
print(num1, num2, num3)

9 9 1


---
## Important Aside: Even or Odd?

Is 6 even or odd?

Is 7 even or odd?

How do you think a computer knows?

Modulus operator shows the remainder of a division problem.

Modding by 2 only gives a `0` or a `1`.

- **4 % 2**:
  - `4 % 2 = 0`. Even!
- **5 % 2**:
  - `5 % 2 = 1`. Odd!
---
## Quick Review

- A variable is a value that can be defined, declared, called and changed within your program.
    - `my_number = 5`
- Naming:
    - Variable names are case sensitive.
    - Use `snake_case`!
- Variables can be reassigned as often as you like, but only the most recent declaration counts.
- Python can do math using operators, such as `+`, `-`, `*`, and `/`
    - You can shorthand the math assignments: `my_num += 7`
---
## Taking a Breather

That was a lot of math!

When it comes down to it, computers operate with a simple, straightforward logic.

Let's switch gears. Up next: Strings!

---
## Introducing Strings

A *character* is:

- Anything on your keyboard , such as a letter or a number.
- "Apple" is five characters: a, p, p, l, e.
- Spaces count! (they're on the keyboard!)

A *string* is:

- A complete list of characters.
- "Apple"
- "Chocolate Cupcake"
- This entire sentence: "Hello, you are 1 of a kind!"

---

## How Do I Create Strings in Python?


You tell Python that your variable will hold a string using quotation marks.

In [17]:
box_contents = "cupcakes" # This is a string
print(box_contents) # It's a normal variable - we can print it.
best_snack = "Frosted Cupcakes" # This is a string.
cupcakes_ive_eaten = 5 # No quotes - this is a number.
cupcakes_ive_eaten_as_string = "5" # Because this is in quotes, this is a string.


cupcakes


---
## We Do: Declaring Strings

A "We Do" means let's practice together. Follow along!

In [18]:
# 1. We'll declare a variable called `name` and assign it the value `Marty`
name = 'Marty'
# 2. We'll declare a variable called `car` and assign it the value `Delorean`
car = 'Delorean'
# 3. We'll declare a variable called `speed` and assign it the *string* value `"88"`
speed = '88'
# 4. We'll print out these variables
print(name, car, speed)
# 5. We'll add `4` to `speed`- what happens?
speed += 4 

Marty Delorean 88


TypeError: can only concatenate str (not "int") to str

---
## String Concatenation

`+` on:

- Numerical variables adds (`5 + 5 = 10`).
- String variables *concatenate* (`"Doc" + "Brown" = "DocBrown"`).
  - *Pssst: Pronunciation tip: con-CAT-en-ATE*
- Numerical strings concatenate to new strings! (`"5" + `"4"` = `"54"`)

In [22]:
first_name = "Doc"
last_name = "Brown"
full_name = first_name + last_name
print (full_name)
# Prints "DocBrown".

DocBrown


---
## We Do: Spaces in Concatenation

It's another "We Do." Let's do this together -  follow along!

In [23]:
sentence = name + "is driving his" + car + speed
print(sentence)

Martyis driving hisDelorean88


We expect the sentence to be `Marty is driving his Delorean 88mph`. Is that what we got?

- Python put the strings together, but do you notice anything wrong? There is no space between the words! This is because we didn't add the spaces in. It's just one of many reasons why we have to carefully watch our spacing and grammar!"
- Since a space is a character - it's on the keyboard - we can make it a string. Therefore, we can add it into our concatenation. By default, concatenation doesn't have spaces - you'll always have to add them yourself.
- You can also `print` directly; you don't necessarily need an extra variable. To print strings next to each other, you separate them with a comma. Then, Python will add the space for you. This isn't concatenating variables, but it's useful to know! Change code to: `sentence = name + " is driving his " + car + " " + speed`
- When `print`ing, commas also create spaces. Change code to: `print(name, "is driving his", car, speed)`
---
## f-strings

- f-strings are a way that you can include variables directly in your strings, without having to use concatenation.
- They are so-called because you create them by adding the letter 'f' just before the string: `f"This is an f-string"`.
- You can then include variables or other Python expressions inside your f-string by wrapping them in `{}`: `print(f'{name} is driving their {car} {speed}mph.')`

In [24]:
print(f'{name} is driving their {car} {speed}mph.')

Marty is driving their Delorean 88mph.


</aside>

---

## Strings and Printing: Review

Strings are made with quotes:



In [25]:
name = "Marty"
car = "Delorean"
speed = "88"



String Concatenation - we need to add the spaces!



In [26]:
sentence = name + " is driving his " + car + " " + speed
string_numbers = "88" + "51"
# string_numbers = 8851



To easily create spaces while printing:



In [27]:
print(name, "is driving his", car, speed)

# Or using f-strings

print(f'{name} is driving their {car} {speed}mph.')

Marty is driving his Delorean 88
Marty is driving their Delorean 88mph.



---

## Discussion: Some Common Mistakes: 1


Do you think this will run? If yes, what does it print?



In [29]:
new_num
print(new_num)

NameError: name 'new_num' is not defined


---

## Discussion: Some Common Mistakes: 2

How about this? Does it run? If so, what does it print?



In [30]:
my_num = 5
print()



<aside class="notes">

Answer: It will run, but it won't print anything.

</aside>

---

## Discussion: Some Common Mistakes: 3

How about this? Does it run? If so, what does it print?



In [31]:
my_num = 5
my_string = "Hello"
print(my_num + my_string)


TypeError: unsupported operand type(s) for +: 'int' and 'str'


<aside class="notes">

Answer: This won't run, because Python will try to add a string and a number and throw an error.

</aside>

---

## Discussion: Some Common Mistakes: 4

One last question. What does this do?



In [32]:
my_num1 = "10"
my_num2 = "20"
print(my_num1 + my_num2)


1020



<aside class="notes">

Answer: There's nothing wrong with this one, except that it makes a string and not a number. It prints the string 1020.

</aside>

---

## Q&A and Summary

We learned a lot today!

- We created, used, and re-assigned number and string variables.
- We used the numerical operators `+ - / * // %`
- We did some complex stuff with the `print` function!

Congrats! You've finished your first programming lesson!

<aside class="notes">

**Teaching Tips**:

- Summarize the lesson and provide a preview of what’s coming next.
- Open your own blank [repl.it](https://repl.it/@GAcoding/blank-repl) in a new tab if needed to recap and be sure everyone is clear.

</aside>


---

## Additional Resources

* [A Repl.it Summarizing Print Statements](https://repl.it/@brandiw/Python-01-Variables-4?lite=true)
* [Python For Beginners](http://www.pythonforbeginners.com/basics/python-variables)
* [Python Programming Tutorial: Variables](https://www.youtube.com/watch?v=vKqVnr0BEJQ)
* [Variables in Python](https://www.guru99.com/variables-in-python.html)
* [Operators Cheatsheet](http://python-reference.readthedocs.io/en/latest/docs/operators/)
* [Python Style Guide: Naming](https://www.python.org/dev/peps/pep-0008/#descriptive-naming-styles)
- [Python-Strings](https://www.tutorialspoint.com/python/python_strings.htm)
- [String Concatenation and Formatting](http://www.pythonforbeginners.com/concatenation/string-concatenation-and-formatting-in-python)
- [String Concatenation and Formatting - Video](https://www.youtube.com/watch?v=jA5LW3bR0Us)
